tests/unit/schema/test_api.py

```diff
@@ -2745,11 +2745,813 @@ def test_read_empty_python_settings_from_service_yaml():
+                    name='FooService',
+                    method=(
+                        descriptor_pb2.MethodDescriptorProto(
+                            name='GetFoo',
+                            input_type='google.example.v1.GetFooRequest',
+                            output_type='google.example.v1.GetFooResponse',
+                        ),
+                        descriptor_pb2.MethodDescriptorProto(
+                            name='GetBar',
+                            input_type='google.example.v1.GetBarRequest',
+                            output_type='google.example.v1.GetBarResponse',
+                        ),
+                    ),
+                ),
+            ),
+        ),
+    )
+
+    service_yaml_config = get_service_yaml_for_selective_gapic_tests(
+        apis=['google.example.v1.FooService'],
+        methods=['google.example.v1.FooService.GetFoo']
+    )
+    opts = Options(service_yaml_config=service_yaml_config)
+
+    api_schema = api.API.build(fd, 'google.example.v1', opts=opts)
+
+    assert 'google.example.v1.Baz' in api_schema.messages
+    assert 'google.example.v1.Foo.FooStatus' in api_schema.enums
+    assert 'google.example.v1.Foo.Bar' in api_schema.messages
+
+    # Establish that we can truncate nested types as well
+    assert 'google.example.v1.Spam' not in api_schema.messages
+    assert 'google.example.v1.Spam.SpamStatus' not in api_schema.enums
+    assert 'google.example.v1.Spam.Ham' not in api_schema.messages
+
+
+@pytest.mark.parametrize("reference_attr", ["type", "child_type"])
+def test_selective_gapic_api_build_with_resources(reference_attr):
+    test_input_names = [
+        ('foo.bar/Foo', 'Foo', 'FooDep', 'GetFooRequest', 'GetFooResponse'),
+        ('foo.bar/Bar', 'Bar', 'BarDep', 'GetBarRequest', 'GetBarResponse'),
+    ]
+
+    messages = []
+
+    for (
+        resource_type,
+        message_name,
+        message_dep_name,
+        request_message_name,
+        response_message_name,
+    ) in test_input_names:
+        resource_message_dep = make_message_pb2(name=message_dep_name)
+
+        # Make sure that we traverse down the fields in the referenced message type.
+        resource_message = make_message_pb2(
+            name=message_name,
+            fields=(
+                make_field_pb2(
+                    name="dep", number=1, type_name=f".google.example.v1.{message_dep_name}"),
+            ),
+        )
+        request_message = make_message_pb2(
+            name=request_message_name,
+            fields=(
+                make_field_pb2(name="thing", number=1, type=9),
+            ),
+        )
+        response_message = make_message_pb2(name=response_message_name)
+
+        # Set up the resource
+        resource_message_opts = resource_message.options.Extensions[resource_pb2.resource]
+        resource_message_opts.type = resource_type
+        resource_message_opts.pattern.append(
+            "foo/{foo}/bar/{bar}")
```

Comment:
Am I correct in thinking these `foo` and `bar` are not required to be the same as the various `Foo` and `Bar` messages and fields you declare? If so, use some other placeholders, like library terms or animal terms.